src/test/munit/headers/validation/headers-validation-test-suite.xml

```diff
@@ -295,4 +295,341 @@
+        <http:headers>#[
+          output application/json
+          ---
+          {
+          'array-header': '{ invalid object, { another invalid object'
+          }]
+        </http:headers>
+        <http:response-validator>
+          <http:success-status-code-validator values="1..500"/>
+        </http:response-validator>
+      </http:request>
+    </munit:execution>
+    <munit:validation>
+      <munit-tools:assert-that expression="#[attributes.statusCode]" is="#[MunitTools::equalTo(400)]"/>
+      <munit-tools:assert-that expression="#[payload.description]" is="#[MunitTools::containsString('Invalid value')]"/>
+    </munit:validation>
+  </munit:test>
+
+  <munit:test name="success-with-string-array-header">
+    <munit:enable-flow-sources>
+      <munit:enable-flow-source value="headers-validation-main"/>
+      <munit:enable-flow-source value="get:\test-array-string-header:headers-validation-config"/>
+    </munit:enable-flow-sources>
+    <munit:execution>
+      <http:request method="GET" config-ref="http-requester-simple" path="/headers/validation/test-array-string-header">
+        <http:headers>#[
+          output application/json
+          ---
+          {
+          'array-header': 'first',
+          'array-header': 'second,third'
+          }]
+        </http:headers>
+        <http:response-validator>
+          <http:success-status-code-validator values="1..500"/>
+        </http:response-validator>
+      </http:request>
+    </munit:execution>
+    <munit:validation>
+      <munit-tools:assert-that expression="#[attributes.statusCode]" is="#[MunitTools::equalTo(200)]"/>
+      <munit-tools:retrieve key="first-header"/>
+      <munit-tools:assert-that expression="#[payload]" is="#[MunitTools::containsString('first')]"/>
+      <munit-tools:retrieve key="second-header"/>
+      <munit-tools:assert-that expression="#[payload]" is="#[MunitTools::containsString('second')]"/>
+    </munit:validation>
+  </munit:test>
+
+  <munit:test name="success-with-csv-string-array-header">
+    <munit:enable-flow-sources>
+      <munit:enable-flow-source value="headers-validation-main"/>
+      <munit:enable-flow-source value="get:\test-array-string-header:headers-validation-config"/>
+    </munit:enable-flow-sources>
+    <munit:execution>
+      <http:request method="GET" config-ref="http-requester-simple" path="/headers/validation/test-array-string-header">
+        <http:headers>#[
+          output application/json
+          ---
+          {
+          'array-header': 'first,second,third'
+          }]
+        </http:headers>
+        <http:response-validator>
+          <http:success-status-code-validator values="1..500"/>
+        </http:response-validator>
+      </http:request>
+    </munit:execution>
+    <munit:validation>
+      <munit-tools:assert-that expression="#[attributes.statusCode]" is="#[MunitTools::equalTo(200)]"/>
```

Comment:
why not assert "first-header" and "second-header"

Reply:
Because of the way that the server test is built, it stores the whole value `first,second,third` in one variable and none for the second. Also, the assertions are being done for the stored values, I think they have nothing to do with the way that they are being processed internally by the router so I'm going to rethink or restructure them.

Reply:
Solved. I've changed the router to update the headers so they are sent parsed downstream to the flows. Also I updated the test server so that the payload stores the size of the headers array and I added the assertions that you were looking for.